gdal/doc/source/development/rfc/rfc76_ogrpythondrivers.rst

```diff
@@ -0,0 +1,397 @@
+  want to be moved into the main repository, it has to be converted to C++.
+  The rationale for this is that the correctness of the Python code can mostly be
+  checked at runtime, whereas C++ benefits from static analysis (at compile time,
+  and other checkers). In the context of GDAL, this rationale also applies. GDAL
+  drivers are also stress-tested by the OSS Fuzz infrastructure, and that requires
+  them to be written in C++.
+
+- The interface between the C++ and Python code might break between GDAL feature
+  releases. In that case we will increment the expected API version number to
+  avoid loading incompatible Python drivers. We will likely not make any effort
+  to be able to deal with plugins of incompatible (previous) API version.
+
+
+SWIG binding changes
+--------------------
+
+None
+
+Security implications
+---------------------
+
+Similar to the existing native code plugin mechanism of GDAL. If the user
+defines the GDAL_PYTHON_DRIVER_PATH environment variable or GDAL_DRIVER_PATH,
+annd put .py scripts in them (or in {prefix}/lib/gdalplugins/python as a fallback),
+they will be executed.
+
+However, opening a .py file with GDALOpen() or similar mechanisms will not
+lead to its execution, so this is safe for normal GDAL usage.
+
+Performance impact
+------------------
+
+If no .py script exists in the researched location, the performance impact on
+GDALAllRegister() should be within the noise.
+
+Backward compatibility
+----------------------
+
+No backward incompatibility. Only functionality addition.
+
+Documentation
+-------------
+
+A tutorial will be added to explain how to write a Python driver.
+
+Testing
+-------
+
+The gdalautotest suite will be extended with a simple test Python driver, along
+the lines of the above example.
+
+Previous discussions
+--------------------
+
+This topic has been discussed in the past in :
+
+- https://lists.osgeo.org/pipermail/gdal-dev/2017-April/thread.html#46526
+- https://lists.osgeo.org/pipermail/gdal-dev/2018-November/thread.html#49294
+
+Implementation
+--------------
+
+A preliminary implementation is available at in
+https://github.com/rouault/gdal/tree/pythondrivers
+
+https://github.com/OSGeo/gdal/compare/master...rouault:pythondrivers
+
+It will be modified to have common code between it and the existing VRT Python
+pixel function mechanism regarding how the Python interpreter is detected/loaded.
+
+Topics for discussion
+---------------------
+
+* Currently that code is unconditionnally compiled in, similarly to the GDAL
+  VRT Python plugin function. Is there a need to have a compile-time switch
+  to disable this ?
+
+* Should we use __iter__() / __next__() methods in the Layer object to iterate
+  over features ? This is the canonical way in Python to define an iterator.
+  However, this code is to be called mainly by the GDAL C++ code, so Python'icity
+  is not critical. Another subtlety is that OGR layers can have spatial and
```

Comment:
I can be someone will complain why the API is not Pythonic :-)

Reply:
I opted for a Python iterator finally